MaxText/maxtext_utils.py

```diff
@@ -0,0 +1,417 @@
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
+
+# pylint: disable=bare-except, consider-using-generator
+"""Utils that are only interesting to MaxText. """
+
+import jax
+import jax.numpy as jnp
+from jax.sharding import PartitionSpec as P
+from jax.experimental.serialize_executable import deserialize_and_load
+from jax.experimental import mesh_utils
+
+import numpy as np
+
+import pickle
+import functools
+import input_pipeline
+import optax
+
+
+import checkpointing
+
+import max_logging
+import os
+import subprocess
+
+from typing import Tuple
+import json
+import flax
+from flax.training import train_state
+from flax import linen as nn
+from flax.linen import partitioning as nn_partitioning
+
+def get_functional_train_with_signature(train_step, mesh, state_mesh_annotations, model, config):
+  """ Get the shardings (both state and data) for train_step """
+  functional_train = get_functional_train_step(train_step, model, config)
+  data_pspec = P(*config.data_sharding)
+  state_mesh_shardings = jax.tree_map(
+      lambda p: jax.sharding.NamedSharding(mesh, p), state_mesh_annotations)
+  data_sharding = jax.tree_map(
+      lambda p: jax.sharding.NamedSharding(mesh, p), data_pspec)
+  in_shardings = (state_mesh_shardings, data_sharding, None) # State, batch, rng
+  out_shardings = (state_mesh_shardings, None, None) # State, metrics, rng
+  static_argnums = () # We partial out the static argnums of model and config
+  donate_argnums = 0 # This is the index of the state - we allow the compiler to make use of this memory.
+  return functional_train, in_shardings, out_shardings, static_argnums, donate_argnums
+
+def get_functional_train_step(train_step, model, config):
+  return functools.partial(train_step, model, config)
+
+def get_optimizer(config, learning_rate_schedule):
+  """ Create AdamW Optimizer following Llama2's training details, see https://arxiv.org/pdf/2307.09288.pdf section 2.2 """
+  return optax.adamw(
+    learning_rate_schedule,
+    b1=config.adam_b1,
+    b2=config.adam_b2,
+    eps=config.adam_eps,
+    eps_root=config.adam_eps_root,
+    weight_decay=config.adam_weight_decay,
+  )
+
+def init_train_state(model, tx, config, key):
+  """
+  We pass in "static" objects like model, tx, config as JAX compares them by
+  object hash, and instantiating them inside causes pjit top-level annotations
+  to fail to match as pytree prefixes if we re-instantiate.
+
+  Args: model, tx, config, key
+  """
+  input_shape = (
+      config.global_batch_size_to_load,
+      config.max_target_length
+  )
+  model_vars = model.init({'params': key, 'dropout': key, 'aqt': key},
+                          jnp.ones(input_shape),
+                          jnp.ones(input_shape))
+  state = train_state.TrainState.create(
+      apply_fn=model.apply,
+      params=model_vars['params'],
+      tx=tx)
+  return state
+
+
+def write_metrics_locally(metrics, step, config, file):
```

Comment:
Its possible this should be moved into shared - probably first by changing the signature to not depend on config. This can be done easily at the time it is desired, if ever.